src/experimental/tools/videoThumbnailLoader/init_source_buffer.ts

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Observable,
+  of as observableOf,
+} from "rxjs";
+import {
+  catchError,
+  filter,
+  map,
+  mapTo,
+  mergeMap,
+  shareReplay,
+} from "rxjs/operators";
+import createSegmentLoader from "../../../core/pipelines/segment/create_segment_loader";
+import { QueuedSourceBuffer } from "../../../core/source_buffers";
+import dash from "../../../transports/dash";
+import prepareSourceBuffer from "./prepare_source_buffer";
+import { IContentInfos } from "./types";
+
+const _currentVideoSourceBuffers: WeakMap<HTMLMediaElement,
+                                          QueuedSourceBuffer<Uint8Array>> = new WeakMap();
+const _currentContentInfos: WeakMap<HTMLMediaElement,
+                                    IContentInfos> = new WeakMap();
+
+const segmentLoader = createSegmentLoader(
+  dash({ lowLatencyMode: false }).video.loader,
+  { maxRetry: 0,
+    maxRetryOffline: 0,
+    initialBackoffDelay: 0,
+    maximumBackoffDelay: 0, });
+
+/**
+ * Get current source buffer :
+ * - If already created for current representation, reuse
+ * - If new codecs and/or init URL, create a new one
+ * @param {Object} contentInfos
+ * @param {HTMLVideoElement} element
+ * @returns {Observable}
+ */
+export function initSourceBuffer$(contentInfos: IContentInfos,
+                                  element: HTMLVideoElement
+): Observable<QueuedSourceBuffer<Uint8Array>> {
+  let _sourceBufferObservable$: Observable<QueuedSourceBuffer<Uint8Array>>;
+  const currentVideoSourceBuffer = _currentVideoSourceBuffers.get(element);
+  const currentContentInfos = _currentContentInfos.get(element);
+  if (currentContentInfos !== undefined &&
+      currentVideoSourceBuffer !== undefined) {
+    _sourceBufferObservable$ = observableOf(currentVideoSourceBuffer);
+  } else {
+    _sourceBufferObservable$ =
+      prepareSourceBuffer(element,
+                          contentInfos.representation.getMimeTypeString())
+        .pipe(
+          map((videoSourceBuffer) => {
+            currentVideoSourceBuffer?.dispose();
+            _currentVideoSourceBuffers.set(element, videoSourceBuffer);
+            return videoSourceBuffer;
+          }),
+          catchError(() => {
+            throw new Error("VideoThumbnailLoaderError: Couldn't open media source.");
+          }),
+          shareReplay()
+        );
+  }
+
+  const currentInitURL =
+    currentContentInfos?.representation.index.getInitSegment()?.mediaURLs?.[0];
+  const initSegment = contentInfos.representation.index.getInitSegment();
+  const initURL = initSegment?.mediaURLs?.[0];
+
+  if (currentContentInfos === undefined ||
+      currentInitURL !== initURL) {
+    return _sourceBufferObservable$.pipe(
+      mergeMap((sourceBuffer) => {
+        if (initSegment == null) {
+          throw new Error("No init segment.");
```

Comment:
Why would that be a problem. If `initSegment` is `null`, it technically means that there is no initialization segment.

I don't know if this case exists but it could (maybe mpeg-ts in Safari?)